### DISCUSSION TITLE
📎 Serialization for AST/CST

### DISCUSSION THREAD

**Discussion by NicholasLYang** [rome/tools#2145] [Suggestions]
### Description

This will be very useful for the playground and for inspecting trees in general.

**Comment by MichaReiser** [comment]
There's an existing string serialization format but that's hard to inspect. I believe it could help to be more explicit about what the goal is because it may then differ significantly how we approach this. 

* Serializing the whole AST can be slow for large files and may, therefore, not be desired
* It may be sufficient to instead expose the metadata for each node kind. Kind of a reflection (readonly) API for `SyntaxNode`s.  
* Have you looked into if we could integrate Rome into [astexplorer](https://astexplorer.net)

**Reply by ematipico** [reply]
> Serializing the whole AST can be slow for large files and may, therefore, not be desired

We could add a `feature` and expose the serialization method only if that feature is enabled.

**Comment by ematipico** [comment]
What's the end goal of the feature? Do you have any idea of what you want to try to achieve at the moment?

**Comment by Stupremee** [comment]
What kind of serialization do you want to achieve? Machine readable for storing them in some data format using `serde`, or human readable representation?

**Reply by ematipico** [reply]
My guess is `serde`. I think what we want to achieve is a way "inspect" the source code from a location and vice-versa. Giving a UI with two windows where in one we have the source code and in the other the AST: 
1. by highlighting a section of the source code we should be able to jump to the correct section of the AST;
2. by clicking on AST node, we should be able to highlight the node in the source code;

**Reply by Stupremee** [reply]
If you're looking for adding `serde` support to the CST, I basically implemented this already [here](https://github.com/domenicquirl/cstree/blob/master/src/serde_impls.rs). It's not the exact same CST, but similar enough. I basically serialize all the events and deserialize a tree by building it up again using `GreenNodeBuilder`. 

If you would use this on a Web UI / Playground, you could read the serialised events in JavaScript and then convert them to something interactive on the page.
